tiledb/sm/array_schema/dimension.h

```diff
@@ -66,6 +66,86 @@ class FilterPipeline;
 enum class Compressor : uint8_t;
 enum class Datatype : uint8_t;
 
+template <class Fn, class... Args>
+inline auto execute_callback_with_type(Datatype type, Fn&& f, Args&&... args) {
```

Comment:
* Interface issues
1. This is very much in the spirit of `std::apply`. `apply_with_type` would be a better name.
2.  With `apply`, function arguments go first, so `f` should be the first argument.
3.  The other arguments should be `type, ...args` afterward.

* Source structure
1. This really shouldn't be in `class Dimension` at all. It's only there because we've been pressing that class (poorly) into service as being (on kind of) a type object. It can be a free function just as `std::apply` is.
2. We have `tiledb/type` for these. It hasn't got much use yet.
3. Just call a new header `apply_with_type.h` for now to avoid prematurely general naming.